### PR TITLE
Make `Digest` support `StorageAppend`

### PR DESCRIPTION
**Pull request by bkchr**
This adds support for `StorageAppend` to `Digest`. Digest is just a
wrapper around a `Vec` and we abuse the fact that SCALE does not puts
any special marker into the encoding for structs. So, we can just append
to the encoded Digest. A test is added that ensures, if the `Digest`
format ever changes, we remove this optimization.

### CODE REVIEW

File: frame/system/src/lib.rs [hunk 1]
```diff
@@ -997,9 +997,7 @@ impl<T: Trait> Module<T> {
 	/// - 1 storage mutation (codec `O(D)`).
```
Comment (tomusdrw) [inline frame/system/src/lib.rs:997]: This does not seem true any more?

Comment (bkchr) [reply frame/system/src/lib.rs]: Updated

File: frame/system/src/lib.rs [hunk 2]
```diff
@@ -993,13 +993,10 @@ impl<T: Trait> Module<T> {
 	/// Deposits a log and ensures it matches the block's log data.
 	///
 	/// # <weight>
-	/// - `O(D)` where `D` length of `Digest`
-	/// - 1 storage mutation (codec `O(D)`).
+	/// - 1 storage write
```
Comment (apopiak) [inline frame/system/src/lib.rs:996]: :+1: 
Could you also adjust the weight of the using function(s)?

Comment (apopiak) [inline frame/system/src/lib.rs:996]:
```suggestion
	/// - `O(1)`
	/// - 1 storage write (codec `O(1)`)
```

Comment (bkchr) [reply frame/system/src/lib.rs:996]: Why `O(1)`?

Comment (apopiak) [reply frame/system/src/lib.rs:996]: That's why we're using `append`, right? To make the runtime complexity constant?

`O(1)` means constant runtime complexity and we want to both annotate the DB accesses for functions as well as their complexity.

Comment (apopiak) [reply frame/system/src/lib.rs:996]: before the complexity was `O(D)`, meaning linear

Comment (shawntabrizi) [reply frame/system/src/lib.rs]: I will adjust it

**Pull request merged**